### PR TITLE
PPS: fixes related to ForwardProton::LHCSector

### PR DESCRIPTION
**Pull request by jan-kaspar**
#### PR description:

This PR contains fixes to two rare problems related to ``ForwardProton::LHCSector``.

1) In ProtonReconstructionAlgorithm.cc: rarely, it may happen that the scattering angles (th_x or th_y) are reconstructed large (mis-reconstruction). In this case, an argument to sqrt becomes negative and consequently proton's momentum z component is assigned NAN. The (undefined) sign of p.z is used in the method ``ForwardProton::lhcSector()`` which in turn returns an invalid value. This PR fixes this problem by checking the sqrt argument and if negative using a realistic value for p.z, yielding correct LHC sector identification.

2) In CTPPSProtonReconstructionSimulationValidator.cc: explicitly check whether proton was reconstructed in LHC sector 56 - protons with invalid sector do no longer pass to the "sector 56" category.

The above problems were spotted in the "direct" simulation with an occurrence O(1 in million events). Therefore the impact of these fixes is expected negligible.

#### PR validation:

The following plots compare results before this PR (blue) and after the PR (red dashed). The top one is for "direct" simulation, the bottom is for Run2 data reconstruction. No differences are observed - compatible with the rare nature of the problems being fixed.

For the few failing events isolated in the "direct" simulation, it has been checked that the proposed PR fixes the problem.

dirsim_cmp
reco_cmp

### CODE REVIEW

File: RecoPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc [hunk 1]
```diff
@@ -389,7 +389,8 @@ reco::ForwardProton ProtonReconstructionAlgorithm::reconstructFromMultiRP(const
   // save reco candidate
   const double sign_z = (armId == 0) ? +1. : -1.;  // CMS convention
   const FP::Point vertex(0., vtx_y, 0.);
-  const double cos_th = sqrt(1. - th_x * th_x - th_y * th_y);
+  const double cos_th_sq = sqrt(1. - th_x * th_x - th_y * th_y);
```
Comment (perrotta) [inline RecoPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc:392]: remove `sqrt`

Comment (jan-kaspar) [reply RecoPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc]: Done in 3ab6103. Many thanks for catching it!

**Comment by perrotta** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
@jan-kaspar do you have an example for the cases in which in the ProtonReconstructionAlgorithm.cc the scattering angles are badly reconstructed? Is it a rounding approximation issue, or really a problem in the reconstruction? In the second case, it may be worth to check and fix the algorithm (while the protection proposed here still remains valid)

**Comment by jan-kaspar** [comment]
> @jan-kaspar do you have an example for the cases in which in the ProtonReconstructionAlgorithm.cc the scattering angles are badly reconstructed? Is it a rounding approximation issue, or really a problem in the reconstruction? In the second case, it may be worth to check and fix the algorithm (while the protection proposed here still remains valid)

Let me describe the origin. Let me focus on single-RP reco - in principle in may occur also for multi-RP reco, but less obviously. The vertical track position in a RP is given by
  y = L(xi) * th_y + v(xi) * vertex_y + effects of sensor resolution + effects misalignment + ... 
Only the first term is considered for reco:
https://github.com/cms-sw/cmssw/blob/master/RecoPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc#L421-L422
the other terms are generally small, O(10 um). However, for some xi values, L(xi) becomes small (it even crosses zero) and therefore the impact of the neglected terms becomes amplified, sometimes leading to obviously unphysical th_y values. In other words, we are dealing with a part of the phase space where we have very low sensitivity to the th_y parameter. This should be reflected by the estimate of the uncertainty:
https://github.com/cms-sw/cmssw/blob/master/RecoPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc#L430
I don't think this can be improved in any way.

**Comment by perrotta** [comment]
Thank you @jan-kaspar 

Since I noticed it in the piece of code that you pointed out to me: maybe we can profit of this PR to (tinily...) improve code perf by avoiding computing a square root at [line 430](https://github.com/cms-sw/cmssw/blob/master/RecoPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc#L430), since it has only to be squared again later on at [line 446](https://github.com/cms-sw/cmssw/blob/master/RecoPPS/ProtonReconstruction/src/ProtonReconstructionAlgorithm.cc#L446)

Please let me know if you intend to also apply that change here, or not.

**Comment by jan-kaspar** [comment]
> Please let me know if you intend to also apply that change here, or not.

Done in ab2e35a

**Comment by perrotta** [comment]
please test
(thank you @jan-kaspar )

**Comment by jfernan2** [comment]
+1

**Comment by perrotta** [comment]
+1
- Fix for pathological rare case, as detailed in the PR description
- Jenkins tests pass

**Comment by silviodonato** [comment]
+1

**Pull request merged**